File: src/dexter/ir/array/StoreExpr.java
package dexter.ir.array;

import dexter.ir.Expr;
import dexter.ir.Visitor;
import dexter.ir.bool.CallExpr;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Maaz Ahmad on 6/25/19.
 */
public class StoreExpr extends CallExpr
{
  public StoreExpr (Expr array, Expr value, List<Expr> indices)
  {
    super("store", new ArrayList<Expr>() {{ add(array); add(value); addAll(indices); }});
  }

  public Expr array() { return args.get(0); }

  public Expr value() { return args.get(1); }

  public List<Expr> index() { return args.subList(2, args.size()); }

  @Override public <T> T accept(Visitor<T> p) { return p.visit(this); }
}

File: src/dexter/frontend/clang/src/ir/array/StoreExpr.cpp
 /*
 * StoreExpr.cpp
 *
 *  Created on: Mar 24, 2018
 *      Author: maazsaf
 */

#include "ir/array/StoreExpr.h"

Dexter::StoreExpr::StoreExpr (Dexter::Expr * array, Dexter::Expr * value, std::vector<Dexter::Expr *> indices) : Dexter::Expr(NULL)
{
	JNIEnv * e = Util::env;

  jclass c = e->FindClass("java/util/ArrayList");
  jmethodID m = e->GetMethodID(c, "<init>", "()V");
  jobject list = e->NewObject(c, m);

  m = e->GetMethodID(c, "add", "(Ljava/lang/Object;)Z");
  std::vector<Expr *>::const_iterator sit;
  for (sit = indices.begin(); sit != indices.end(); ++sit)
    e->CallObjectMethod(list, m, (*sit)->obj());

  c = e->FindClass("dexter/ir/array/StoreExpr");
  m = e->GetMethodID(c, "<init>", "(Ldexter/ir/Expr;Ldexter/ir/Expr;Ljava/util/List;)V");
  this->mObj = e->NewObject(c, m, array->obj(), value->obj(), list);
}